File: python/jittor/src/lock.cc
// ***************************************************************
// Copyright (c) 2023 Jittor. All Rights Reserved.
// Maintainers:
//     Wenyang Zhou <576825820@qq.com>
//     Dun Liang <randonlang@gmail.com>
//
// This file is subject to the terms and conditions defined in
// file 'LICENSE.txt', which is part of this source code package.
// ***************************************************************
#include <stdio.h>
#ifdef _WIN32
#include <windows.h>
#include <fileapi.h>
#include <process.h>
#include <io.h>
#define getpid _getpid
#define open _open
#else
#include <unistd.h>
#include <sys/file.h>
#endif
#include <fcntl.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <chrono>
#include <thread>

#include "lock.h"

namespace jittor {

// The descriptor is opened by jittor_utils/lock.py and handed over here, so
// that both languages hold one lock of one kind on one open file description.
//
// This used to be a second open() of the same path plus fcntl(F_SETLKW), a
// POSIX record lock. On Linux a record lock and the flock() the Python side
// takes are independent lock families: neither excludes the other, so both
// sides could be "holding jittor.lock" at once and compile into the same cache
// directory. Record locks also have the rule that closing *any* descriptor for
// the file releases all of the process's record locks on it, which meant a
// Python lock object being garbage-collected quietly released this one.
#ifdef _WIN32
static HANDLE lock_handle = INVALID_HANDLE_VALUE;
#endif
static int lock_fd = -1;
int _has_lock = 0;

DEFINE_FLAG(bool, disable_lock, 0, "Disable file lock");

static double env_seconds(const char* name, double _default) {
    const char* v = getenv(name);
    if (!v || !v[0]) return _default;
    char* end = nullptr;
    double parsed = strtod(v, &end);
    if (end == v) {
        LOGw << name << "=" << v << "is not a number, using" << _default;
        return _default;
    }
    return parsed;
}

// Keep these in sync with jittor_utils/lock.py, which documents them.
static double lock_timeout() {
    static double v = env_seconds("JT_LOCK_TIMEOUT", 1800.0);
    return v;
}
static double lock_report_after() {
    static double v = env_seconds("JT_LOCK_REPORT_AFTER", 30.0);
    return v;
}

static string self_cmdline() {
    string cmd;
#ifndef _WIN32
    FILE* f = fopen("/proc/self/cmdline", "rb");
    if (f) {
        char buf[512];
        size_t n = fread(buf, 1, sizeof(buf), f);
        fclose(f);
        for (size_t i=0; i<n; i++)
            cmd += buf[i] ? buf[i] : ' ';
    }
#endif
    return cmd;
}

// Same JSON record shape the Python side writes, so either can read the other's.
static void write_holder() {
#ifndef _WIN32
    if (lock_fd < 0) return;
    auto now = std::chrono::duration_cast<std::chrono::seconds>(
        std::chrono::system_clock::now().time_since_epoch()).count();
    string record = string("{\"pid\": ") + std::to_string(getpid())
        + ", \"time\": " + std::to_string((int64)now)
        + ", \"cmd\": \"" + self_cmdline() + "\"}\n";
    if (ftruncate(lock_fd, 0) != 0) return;
    ssize_t written = pwrite(lock_fd, record.c_str(), record.size(), 0);
    (void)written;
#endif
}

static string describe_holder() {
#ifndef _WIN32
    if (lock_fd < 0) return "holder unknown";
    char buf[4096];
    ssize_t n = pread(lock_fd, buf, sizeof(buf)-1, 0);
    if (n <= 0) return "holder unknown (no record was written)";
    buf[n] = 0;
    return string("holder record: ") + buf;
#else
    return "holder unknown";
#endif
}

void set_lock_fd(int64 fd, bool has_lock) {
#ifdef _WIN32
    lock_handle = (HANDLE)fd;
#else
    lock_fd = (int)fd;
#endif
    _has_lock = has_lock ? 1 : 0;
    LOGv << "SHARE LOCK fd:" << fd << "has_lock:" << _has_lock
         << "Pid:" << getpid();
}

void lock() {
    if (disable_lock) return;
#ifdef _WIN32
    ASSERT(lock_handle != INVALID_HANDLE_VALUE);
    OVERLAPPED offset = {0, 0, 0, 0, NULL};
    ASSERT(LockFileEx(lock_handle, 2, 0, -0x10000, 0, &offset));
#else
    ASSERT(lock_fd >= 0);
    auto start = std::chrono::steady_clock::now();
    bool reported = false;
    while (1) {
        if (flock(lock_fd, LOCK_EX | LOCK_NB) == 0) break;
        if (errno != EWOULDBLOCK && errno != EINTR)
            LOGf << "could not lock the build lock:" << strerror(errno)
                 << ". Set disable_lock=1 to build without it (unsafe if"
                 << "anything else builds into the same cache).";
        double waited = std::chrono::duration<double>(
            std::chrono::steady_clock::now() - start).count();
        if (!reported && waited >= lock_report_after()) {
            LOGw << "waiting for the build lock," << describe_holder();
            reported = true;
        }
        if (lock_timeout() > 0 && waited >= lock_timeout())
            LOGf << "timed out after" << waited
                 << "s waiting for the build lock," << describe_holder()
                 << ". Raise JT_LOCK_TIMEOUT (0 waits forever) if a cold"
                 << "build really takes this long.";
        std::this_thread::sleep_for(std::chrono::milliseconds(
            waited < 1 ? 50 : 500));
    }
    write_holder();
#endif
    _has_lock = 1;
    LOGvv << "LOCK Pid:" << getpid();
}

void unlock() {
    if (disable_lock) return;
#ifdef _WIN32
    ASSERT(lock_handle != INVALID_HANDLE_VALUE);
    OVERLAPPED offset = {0, 0, 0, 0, NULL};
    ASSERT(UnlockFileEx(lock_handle, 0, -0x10000, 0, &offset));
#else
    ASSERT(lock_fd >= 0);
    ASSERT(flock(lock_fd, LOCK_UN) == 0);
#endif
    _has_lock = 0;
    LOGvv << "UNLOCK Pid:" << getpid();
}

void lock_acquire() {
    if (_has_lock) return;
    lock();
}

void lock_release() {
    if (!_has_lock) return;
    unlock();
}

bool lock_is_held() {
    return _has_lock != 0;
}

} // jittor

File: python/jittor_utils/lock.py
# ***************************************************************
# Copyright (c) 2023 Jittor. All Rights Reserved.
# Maintainers: Dun Liang <randonlang@gmail.com>.
# This file is subject to the terms and conditions defined in
# file 'LICENSE.txt', which is part of this source code package.
# ***************************************************************
'''The single build lock shared by the Python and the C++ side.

Both sides of Jittor serialize compilation against ``jittor.lock``. They used
to do it with two *different* kinds of POSIX lock on two *different* file
descriptors: Python called ``fcntl.flock`` (a BSD lock, attached to the open
file description) while C++ called ``fcntl(F_SETLKW)`` (a POSIX record lock,
attached to the (process, inode) pair). On Linux those two lock families are
completely independent -- taking one never blocks the other -- so a Python
download and a C++ op compilation could write the same cache directory at the
same time while each believed it held the lock. The record lock brought a
second trap along: closing *any* descriptor for a file drops *all* of that
process's record locks on it, so a garbage-collected Python lock object could
silently release the lock C++ was relying on.

This module is now the only implementation. There is exactly one descriptor,
opened here, and exactly one kind of lock (``flock``) taken on it. Once
``jittor_core`` is loaded the descriptor and the "do we hold it" flag are
handed to C++ (:meth:`Lock.bind_core`), and every later lock/unlock from
either language goes through that single shared state.
'''

import errno
import json
import os
import sys
import time

try:
    import fcntl
except ImportError:
    fcntl = None
    try:
        import msvcrt
        import win32file
        import pywintypes
        _OVERLAPPED = pywintypes.OVERLAPPED()
    except:
        raise Exception("""pywin32 package not found, please install it.
>>> python3.x -m pip install pywin32
If conda is used, please install with command:
>>> conda install pywin32""")

from jittor_utils import cache_path, LOG

disable_lock = os.environ.get("disable_lock", "0") == "1"


def _env_seconds(name, default):
    value = os.environ.get(name)
    if value is None or value == "":
        return default
    try:
        return float(value)
    except ValueError:
        LOG.w(f"{name}={value!r} is not a number, using {default}")
        return default


# How long to wait for jittor.lock before giving up. A cold build of the core
# takes minutes and a second process legitimately waits for all of it, so the
# default is generous; 0 means wait forever (the old behaviour).
lock_timeout = _env_seconds("JT_LOCK_TIMEOUT", 1800.0)
# How long to wait before naming the current holder. An orphaned or wedged
# holder used to produce no output at all, for as long as it took.
lock_report_after = _env_seconds("JT_LOCK_REPORT_AFTER", 30.0)

_HOLDER_RECORD_SIZE = 4096


def _self_cmdline():
    try:
        with open("/proc/self/cmdline", "rb") as f:
            argv = f.read().split(b"\0")
    except OSError:
        argv = [a.encode("utf8", "replace") for a in sys.argv]
    cmd = " ".join(a.decode("utf8", "replace") for a in argv if a)
    return cmd[:512]


def _write_holder(fd):
    '''Record who holds the lock, so a waiter can name the process it waits for.'''
    record = json.dumps({
        "pid": os.getpid(),
        "time": time.time(),
        "cmd": _self_cmdline(),
    })
    blob = record.encode("utf8")[:_HOLDER_RECORD_SIZE - 1] + b"\n"
    try:
        os.ftruncate(fd, 0)
        os.pwrite(fd, blob, 0)
    except OSError as e:
        # A read-only or full filesystem must not stop us from building; the
        # only thing lost is the diagnostic.
        LOG.v(f"could not record lock holder: {e}")


def _pid_alive(pid):
    if os.path.isdir("/proc"):
        return os.path.isdir(f"/proc/{pid}")
    try:
        os.kill(pid, 0)
    except OSError as e:
        return e.errno != errno.ESRCH
    return True


def describe_holder(fd):
    '''Human readable description of whoever wrote the holder record.'''
    try:
        raw = os.pread(fd, _HOLDER_RECORD_SIZE, 0)
    except (OSError, AttributeError):
        return "holder unknown"
    text = raw.split(b"\0", 1)[0].decode("utf8", "replace").strip()
    if not text:
        return "holder unknown (no record was written)"
    try:
        record = json.loads(text)
        pid = int(record["pid"])
    except (ValueError, KeyError, TypeError):
        return f"holder record: {text[:200]}"
    held_for = ""
    if isinstance(record.get("time"), (int, float)):
        held_for = f" for {time.time() - record['time']:.0f}s"
    alive = "" if _pid_alive(pid) else \
        " -- that process is gone, the record is stale"
    return f"held by pid {pid}{held_for}: {record.get('cmd', '')}{alive}"


def _acquire(fd, path):
    '''Take the exclusive lock on ``fd``, reporting and timing out while waiting.

    ``flock`` has no timeout, so this polls with ``LOCK_NB`` instead of using a
    blocking call. Polling costs a wakeup every half second while waiting,
    which is nothing next to the compile it is waiting for, and buys the two
    things a blocking wait could never give: a message naming the holder, and
    an error instead of an unbounded hang.
    '''
    if fcntl is None:
        hfile = win32file._get_osfhandle(fd)
        win32file.LockFileEx(hfile, 2, 0, -0x10000, _OVERLAPPED)
        return
    start = time.monotonic()
    reported = False
    while True:
        try:
            fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
            return
        except OSError as e:
            if e.errno not in (errno.EACCES, errno.EAGAIN, errno.EWOULDBLOCK):
                raise RuntimeError(
                    f"could not lock {path}: {e}. Set disable_lock=1 to run "
                    f"without the build lock (unsafe if anything else is "
                    f"building into the same cache).") from e
        waited = time.monotonic() - start
        if not reported and waited >= lock_report_after:
            LOG.w(f"waiting for build lock {path}, {describe_holder(fd)}")
            reported = True
        if lock_timeout > 0 and waited >= lock_timeout:
            raise RuntimeError(
                f"timed out after {waited:.0f}s waiting for the build lock "
                f"{path}, {describe_holder(fd)}. Raise JT_LOCK_TIMEOUT (0 "
                f"waits forever) if a cold build really takes this long.")
        time.sleep(0.05 if waited < 1 else 0.5)


def _release(fd):
    if fcntl is None:
        hfile = win32file._get_osfhandle(fd)
        win32file.UnlockFileEx(hfile, 0, -0x10000, _OVERLAPPED)
        return
    fcntl.flock(fd, fcntl.LOCK_UN)


class Lock:
    '''Exclusive lock on one file, shared with the C++ side once it is loaded.'''

    def __init__(self, filename):
        self.filename = filename
        # O_CREAT rather than truncating: the body of the file is the holder
        # record, and a waiter must be able to read the holder's record rather
        # than an empty file left behind by its own open().
        self.fd = os.open(filename, os.O_RDWR | os.O_CREAT, 0o666)
        LOG.v(f'OPEN LOCK path: {filename} PID: {os.getpid()}')
        self._py_is_locked = False
        # jittor_core, once compiler.py has handed it this descriptor. From
        # then on it owns the state and both languages go through it.
        self.core = None

    def fileno(self):
        return self.fd

    def bind_core(self, core):
        '''Hand the descriptor and the current state to the C++ side.

        Called once, right after ``jittor_core`` is imported. Everything before
        this point ran with no C++ code loaded at all, so there was no second
        holder to disagree with.
        '''
        handle = self.fd
        if fcntl is None:
            handle = msvcrt.get_osfhandle(self.fd)
        core.set_lock_fd(handle, self._py_is_locked)
        self.core = core

    @property
    def is_locked(self):
        if self.core is not None:
            return self.core.lock_is_held()
        return self._py_is_locked

    def lock(self):
        if disable_lock:
            return
        if self.core is not None:
            self.core.lock_acquire()
            return
        if self._py_is_locked:
            return
        _acquire(self.fd, self.filename)
        _write_holder(self.fd)
        self._py_is_locked = True
        LOG.vv(f'LOCK PID: {os.getpid()}')

    def unlock(self):
        if disable_lock:
            return
        if self.core is not None:
            self.core.lock_release()
            return
        if not self._py_is_locked:
            return
        _release(self.fd)
        self._py_is_locked = False
        LOG.vv(f'UNLOCK PID: {os.getpid()}')

    # NOTE: deliberately no __del__. The descriptor lives as long as the
    # process and the kernel closes it at exit. Closing it from a finalizer is
    # what used to drop the C++ side's lock out from under it.


class _base_scope:
    '''base_scope for support @xxx syntax'''
    def __enter__(self): pass
    def __exit__(self, *exc): pass
    def __call__(self, func):
        def inner(*args, **kw):
            with self:
                ret = func(*args, **kw)
            return ret
        return inner

class lock_scope(_base_scope):
    def __enter__(self):
        self.is_locked = jittor_lock.is_locked
        if not self.is_locked:
            jittor_lock.lock()

    def __exit__(self, *exc):
        if not self.is_locked:
            jittor_lock.unlock()

class unlock_scope(_base_scope):
    def __enter__(self):
        self.is_locked = jittor_lock.is_locked
        if self.is_locked:
            jittor_lock.unlock()

    def __exit__(self, *exc):
        if self.is_locked:
            jittor_lock.lock()

lock_path = os.path.abspath(os.path.join(cache_path, "../jittor.lock"))
jittor_lock = Lock(lock_path)
